File: rxjava/src/main/java/xyz/truenight/databinding/rxjava/RxObservableField.java
package xyz.truenight.databinding.rxjava;

import android.databinding.ObservableField;

import java.util.Set;
import java.util.concurrent.CopyOnWriteArraySet;

import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;
import io.reactivex.subjects.BehaviorSubject;
import xyz.truenight.utils.Optional;

/**
 * Copyright (C) 2017 Mikhail Frolov
 */

public class RxObservableField<T> extends ObservableField<T> {

    private Set<OnPropertyChangedCallback> count = new CopyOnWriteArraySet<>();

    private Disposable subscription;

    private final Observable<Optional<T>> observable;

    public RxObservableField(Observable<Optional<T>> observable) {
        super();
        this.observable = observable;
    }

    public RxObservableField(T value, Observable<Optional<T>> observable) {
        super(value);
        this.observable = observable;
    }

    public RxObservableField(BehaviorSubject<Optional<T>> subject) {
        super(safe(subject.getValue()));
        this.observable = subject;
    }

    private static <T> T safe(Optional<T> value) {
        return value == null ? null : value.orElse(null);
    }

    private void safeSet(Optional<T> value) {
        super.set(safe(value));
    }

    @Deprecated
    @Override
    public void set(T value) {
        super.set(value);
    }

    @Override
    public void addOnPropertyChangedCallback(OnPropertyChangedCallback callback) {
        super.addOnPropertyChangedCallback(callback);
        if (count.isEmpty()) {
            subscription = observable.subscribe(this::safeSet);
        }
        count.add(callback);
    }

    @Override
    public void removeOnPropertyChangedCallback(OnPropertyChangedCallback callback) {
        super.removeOnPropertyChangedCallback(callback);
        count.remove(callback);
        if (count.isEmpty()) {
            if (subscription != null && !subscription.isDisposed()) {
                subscription.dispose();
            }
        }
    }
}

File: rxjava/src/main/java/xyz/truenight/databinding/rxjava/RxObservableField.kt
package xyz.truenight.databinding.rxjava

import androidx.databinding.ObservableField
import io.reactivex.Flowable
import io.reactivex.Observable
import io.reactivex.disposables.Disposable
import xyz.truenight.utils.optional.Optional
import xyz.truenight.utils.optional.toOptional
import java.util.concurrent.CopyOnWriteArraySet

/**
 * Copyright (C) 2017 Mikhail Frolov
 */

class RxObservableField<T : Any> @JvmOverloads internal constructor(private val observable: Observable<Optional<T>>, value: T? = null, private val setter: ((T?) -> Unit)? = null) : ObservableField<T?>(value) {

    private val count = CopyOnWriteArraySet<androidx.databinding.Observable.OnPropertyChangedCallback>()

    private var subscription: Disposable? = null

    override fun set(value: T?) {
        super.set(value)
        setter?.let { it(value) }
    }

    override fun addOnPropertyChangedCallback(callback: androidx.databinding.Observable.OnPropertyChangedCallback) {
        super.addOnPropertyChangedCallback(callback)
        if (count.isEmpty()) {
            subscription = observable.subscribe { super.set(it.value) }
        }
        count.add(callback)
    }

    override fun removeOnPropertyChangedCallback(callback: androidx.databinding.Observable.OnPropertyChangedCallback) {
        super.removeOnPropertyChangedCallback(callback)
        count.remove(callback)
        if (count.isEmpty()) {
            subscription?.dispose()
        }
    }
}

fun <T : Any> Observable<Optional<T>>.toBinding(default: Optional<T> = Optional.empty(), setter: ((T?) -> Unit)? = null) =
        RxObservableField(this, default.value, setter)

fun <T : Any> Flowable<Optional<T>>.toBinding(default: Optional<T> = Optional.empty(), setter: ((T?) -> Unit)? = null) =
        RxObservableField(this.toObservable(), default.value, setter)

fun <T : Any> Observable<T>.toBinding(default: T? = null, setter: ((T?) -> Unit)? = null) =
        RxObservableField(this.map { it.toOptional() }, default, setter)

fun <T : Any> Flowable<T>.toBinding(default: T? = null, setter: ((T?) -> Unit)? = null) =
        RxObservableField(this.toObservable().map { it.toOptional() }, default, setter)
